From 910b48c7dfff1a84924c674ee530a0d7d42e9ee3 Mon Sep 17 00:00:00 2001
From: Heiko Schocher <hs@denx.de>
Date: Sat, 25 Feb 2023 17:02:56 +0100
Subject: [PATCH 10/21] tbot/machine/board/uboot.py: HACK autoboot_iter

search autoboot_iter times for autoboot string.

until I do not know how to empty a received channel, I have
the problem, using a lauterbach debugger and not empty
flash, that autoboot string occurs before debugger stops
the board, So I need to empty recieve buffer before
I send go command ... no idea yet how to, so this is a HACK

Signed-off-by: Heiko Schocher <hs@denx.de>
---
 tbot/machine/board/uboot.py | 38 +++++++++++++++++++++----------------
 1 file changed, 22 insertions(+), 16 deletions(-)

diff --git a/tbot/machine/board/uboot.py b/tbot/machine/board/uboot.py
index 19cfd07..e79e4cd 100644
--- a/tbot/machine/board/uboot.py
+++ b/tbot/machine/board/uboot.py
@@ -98,6 +98,7 @@ class UBootAutobootIntercept(machine.Initializer, UbootStartup):
     """
     Keys to press as soon as autoboot prompt is detected.
     """
+    autoboot_iter: typing.Union[int] = 1
 
     @property
     @abc.abstractmethod
@@ -115,23 +116,28 @@ class UBootAutobootIntercept(machine.Initializer, UbootStartup):
             self.ch.add_death_string(d)
 
         if self.autoboot_prompt is not None:
-            with self.ch.with_stream(self._uboot_startup_event()):
-                timeout = None
-                if self.boot_timeout is not None:
-                    assert self._timeout_start is not None
-                    timeout = self.boot_timeout - (
-                        time.monotonic() - self._timeout_start
-                    )
+            i = 0
+            while i < self.autoboot_iter:
+                with self.ch.with_stream(self._uboot_startup_event()):
+                    timeout = None
+                    if self.boot_timeout is not None:
+                        assert self._timeout_start is not None
+                        timeout = self.boot_timeout - (
+                            time.monotonic() - self._timeout_start
+                        )
+
+                    try:
+                        self.ch.read_until_prompt(
+                            prompt=self.autoboot_prompt, timeout=timeout
+                        )
+                    except TimeoutError:
+                        raise TimeoutError(
+                            "U-Boot autoboot prompt did not show up in time"
+                        ) from None
+                    i += 1
+                    if i == self.autoboot_iter:
+                        self.ch.send(self.autoboot_keys, _ignore_blacklist=True)
 
-                try:
-                    self.ch.read_until_prompt(
-                        prompt=self.autoboot_prompt, timeout=timeout
-                    )
-                except TimeoutError:
-                    raise TimeoutError(
-                        "U-Boot autoboot prompt did not show up in time"
-                    ) from None
-                self.ch.send(self.autoboot_keys, _ignore_blacklist=True)
 
         yield None
 
-- 
2.31.1

